edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs

```diff
@@ -238,9 +249,57 @@ async void OnConnectionStatusChanged(ConnectionStatus status, ConnectionStatusCh
 
                 if (this.pullOnReconnect && this.initTask.IsCompleted && status == ConnectionStatus.Connected)
                 {
+                    var delayedTwinPull = true;
                     using (await this.twinLock.LockAsync())
                     {
-                        await this.RefreshTwinAsync();
+                        var now = DateTime.Now;
+                        if (now - this.lastTwinPullOnConnect > this.TwinPullOnConnectThrottleTime && !this.isDelayedTwinPullInProgress.Get())
+                        {
+                            this.lastTwinPullOnConnect = now;
+                            await this.RefreshTwinAsync();
+                            delayedTwinPull = false;
+                        }
+                    }
+
+                    if (delayedTwinPull)
+                    {
+                        if (this.isDelayedTwinPullInProgress.GetAndSet(true))
+                        {
+                            this.pullRequestCounter++;
+                        }
+                        else
+                        {
+                            _ = Task.Run(
+                                async () =>
+                                {
+                                    await Task.Delay(this.TwinPullOnConnectThrottleTime);
+
+                                    var requestCounter = default(int);
+                                    using (await this.twinLock.LockAsync())
+                                    {
+                                        this.lastTwinPullOnConnect = DateTime.Now;
+
+                                        try
+                                        {
+                                            await this.RefreshTwinAsync();
+                                        }
+                                        catch
+                                        {
+                                            // swallowing intentionally
```

Comment:
So the expectation is that if the async pull for twin fails, we continue to use whatever the stored twin is?

Reply:
Log debug?

Reply:
It uses the original logic (where it was a RefreshTwinAsync() call by itself. First, I did not even put it into a try block (as RefreshTwinAsync() handles most of the exceptions except the "fatal" ones (like OutOfMemory), but then I though that if I misunderstand the code, or some future change happens, then the "IsDelayedTwinPullInProgress" flag gets stuck and no twin pull would happen afterwards. So then I put it into a finally, but then it looked silly that the rest 3 lines was in a finally block, so then I changed to catch and moved out those three lines out of the finally block - that is why the code looks like this
(also, the logging happens in RefreshTwinAsync() so a log here would just duplicate it)